### PR TITLE
Add random_uniform_distribution_deflected to all features and enable random number seed as input

### PR DESCRIPTION
**Pull request by Wang-yijun**
This PR adds random_uniform_distribution_deflected texture to all features (continental plate, oceanic plate, fault, subducting plate and mantle layer), sets default global random number seed as NaN unless entered in the input file, and lets random_uniform_distribution_deflected texture take random number seeds as inputs to generate different random textures with different deflection from the center.

### CODE REVIEW

File: include/world_builder/world.h [hunk 1]
```diff
@@ -275,6 +275,11 @@ namespace WorldBuilder
        */
       std::string interpolation;
 
+      /**
+       * Todo
+       */
+      int local_seed;
```
Comment (MFraters) [inline include/world_builder/world.h:281]: I don't think this is needed. Can you remove this and just locally declare it as an int?

File: source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        RandomUniformDistributionDeflected::~RandomUniformDistributionDeflected()
+          = default;
+
+        void
+        RandomUniformDistributionDeflected::declare_entries(Parameters &prm, const std::string & /*unused*/)
+        {
+          // Document plugin and require entries if needed.
+          // Add compositions the required parameters.
```
Comment (MFraters) [inline source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc:64]:
```suggestion
          // Add compositions to the required parameters.
```

File: source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+          max_depth_surface = Objects::Surface(prm.get("max depth",coordinates));
+          max_depth = max_depth_surface.maximum;
+          compositions = prm.get_vector<unsigned int>("compositions");
+
+          operation = prm.get<std::string>("orientation operation");
+          grain_sizes = prm.get_vector<double>("grain sizes");
+          normalize_grain_sizes = prm.get_vector<bool>("normalize grain sizes");
+          deflections = prm.get_vector<double>("deflections");
+          seed = prm.get<unsigned int>("random number seed");
+          world->get_random_number_engine().seed(seed);
```
Comment (MFraters) [inline source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc:116]: As discussed, use global one

**Review by MFraters — COMMENTED** [review]
Looks good. Here are some initial comments we also discussed.

File: source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc (hunk 2 above)
Comment (Wang-yijun) [reply source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc]: Fixed this and some other descriptions of the feature in random_uniform_distribution_deflected

File: source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc (hunk 3 above)
Comment (Wang-yijun) [reply source/world_builder/features/continental_plate_models/grains/random_uniform_distribution_deflected.cc]: Deleted this and other use of local seed inside features

File: include/world_builder/world.h [hunk 4]
```diff
@@ -276,6 +276,7 @@ namespace WorldBuilder
       std::string interpolation;
 
 
+
```
Comment (MFraters) [inline include/world_builder/world.h:279]: could you remove this?

**Review by MFraters — COMMENTED** [review]
The code looks good to me. Can you add tests to that all code is tested?

File: include/world_builder/world.h (hunk 4 above)
Comment (Wang-yijun) [reply include/world_builder/world.h]: Test added and empty line removed.

File: include/world_builder/features/fault_models/grains/random_uniform_distribution_deflected.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  namespace Features
+  {
+    using namespace FeatureUtilities;
+    namespace FaultModels
+    {
+      namespace Grains
+      {
+        /**
+         * This class represents a fault and can implement
```
Comment (tjhei) [inline include/world_builder/features/fault_models/grains/random_uniform_distribution_deflected.h:37]: Is this comment block copied from somewhere?

File: include/world_builder/features/mantle_layer_models/grains/random_uniform_distribution_deflected.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            get_grains(const Point<3> &position,
+                       const Objects::NaturalCoordinate &position_in_natural_coordinates,
+                       const double depth,
+                       const unsigned int composition_number,
+                       WorldBuilder::grains grains,
+                       const double feature_min_depth,
+                       const double feature_max_depth) const override final;
+
+          private:
+            // random uniform distribution deflected grains submodule parameters
```
Comment (tjhei) [inline include/world_builder/features/mantle_layer_models/grains/random_uniform_distribution_deflected.h:98]: I would delete this comment as it doesn't add any useful information.

File: include/world_builder/features/fault_models/grains/random_uniform_distribution_deflected.h (hunk 5 above)
Comment (Wang-yijun) [reply include/world_builder/features/fault_models/grains/random_uniform_distribution_deflected.h:37]: Yes, I this is copied from random_uniform_distribution.h in the same folder as a template and I made changes based on that.

File: include/world_builder/features/mantle_layer_models/grains/random_uniform_distribution_deflected.h (hunk 6 above)
Comment (Wang-yijun) [reply include/world_builder/features/mantle_layer_models/grains/random_uniform_distribution_deflected.h]: Removed.

**Review by MFraters — APPROVED** [review]
I would propose merging this as is. Please make a follow up pull request with

1. a more extensive explanation of the random number seed declare entry
2. A test which is exactly the same, but has a different random number.

**Comment by Wang-yijun** [comment]
Example texture generated using random uniform distribution deflected (visualized using MTEX package in Matlab):

deflection = 0.8, random number seed = 1

deflection = 0.6, random number seed = 1

deflection = 0.4, random number seed = 1

deflection = 0.2, random number seed = 1

deflection = 0.6, random number seed = 10

**Pull request merged**